clippy_lints/src/cognitive_complexity.rs

```diff
@@ -48,190 +52,258 @@ impl LintPass for CognitiveComplexity {
+                        "you could split it up into multiple smaller functions",
+                    );
+                }
+            }
         }
     }
 
-    fn enter_lint_attrs(&mut self, cx: &LateContext<'a, 'tcx>, attrs: &'tcx [Attribute]) {
+    fn enter_lint_attrs(&mut self, cx: &EarlyContext<'_>, attrs: &[Attribute]) {
         self.limit.push_attrs(cx.sess(), attrs, "cognitive_complexity");
     }
-    fn exit_lint_attrs(&mut self, cx: &LateContext<'a, 'tcx>, attrs: &'tcx [Attribute]) {
+    fn exit_lint_attrs(&mut self, cx: &EarlyContext<'_>, attrs: &[Attribute]) {
         self.limit.pop_attrs(cx.sess(), attrs, "cognitive_complexity");
     }
 }
 
-struct CCHelper<'a, 'tcx: 'a> {
-    match_arms: u64,
-    divergence: u64,
-    returns: u64,
-    short_circuits: u64, // && and ||
-    cx: &'a LateContext<'a, 'tcx>,
+struct CoCHelper {
+    current_nesting: u64,
+    score: f64,
 }
 
-impl<'a, 'tcx> Visitor<'tcx> for CCHelper<'a, 'tcx> {
-    fn visit_expr(&mut self, e: &'tcx Expr) {
-        match e.node {
-            ExprKind::Match(_, ref arms, _) => {
-                walk_expr(self, e);
-                let arms_n: u64 = arms.iter().map(|arm| arm.pats.len() as u64).sum();
-                if arms_n > 1 {
-                    self.match_arms += arms_n - 2;
-                }
+impl CoCHelper {
+    fn push_nesting(&mut self) {
+        self.current_nesting += 1;
+    }
+
+    fn pop_nesting(&mut self) {
+        assert!(self.current_nesting > 0);
+        self.current_nesting -= 1;
+    }
+
+    fn add_to_score(&mut self, amount: u64) {
+        self.score += amount as f64;
+    }
+
+    fn score_nd_structure(&mut self) {
+        self.add_to_score(self.current_nesting);
+        self.score_ni_structure();
+    }
+
+    fn score_ni_structure(&mut self) {
+        self.add_to_score(1);
+    }
+}
+
+// FIXME: delete these notes
+
+/*
+    Notes on debugging this (14/4):
+    // tests/ui/cognitive_complexity.rs
+    * Check why is 330 not being flagged (it has CoC 0 atm)
+*/
+
+impl<'ast> Visitor<'ast> for CoCHelper {
+    /*
+    # Implemented here:
+
+    ## Nesting Structures
+        IMPORTANT: ExprKind::Block(..)
+        already covers all cases.
+    ## (Nesting-Dependent) Increments
+        if
+        match
+        for, while, loop
+    ## (Nesting-Independent) Increments
+        break, continue
+        Sequences of binary logical operators
+        Recursive function calls
+        Macro calls
+    */
+
+    fn visit_expr(&mut self, ex: &'ast Expr) {
+        match ex.node {
+            // Nesting-Increasing (the one and only)
+            ExprKind::Block(..) => {
+                self.push_nesting();
+                walk_expr(self, ex);
+                self.pop_nesting();
+            },
+
+            ExprKind::Closure(.., _) | ExprKind::IfLet(..) => {
+                // If Let is free of own increment.
+                // This is because it's a language construct
```

Comment:
This may stop being true once I have implemented https://github.com/rust-lang/rust/issues/53667 in the AST such that we'll have `ExprKind::Let(...)`.

Reply:
Oh, please ping me when you do so that I can make the necessary tweaks! ^^

Reply:
I'll try to remember :P Chances are this will require reworking in Clippy in many places.

Reply:
Wooow. If I understood the RFC correctly, you're basically including the `if_chain` macro's semantics in the language. That's huge! I don't know how the ASTs will change after that.

I guess I'll cross that bridge when we get to it ^^

Reply:
@felix91gr Yeah pretty much `if_chain!` in the language but nicer. =)

The current sketch is that on the AST level we'll have `ast::ExprKind::Let(Vec<P<Pat>>, P<Expr>)` and then we have `ast::ExprKind::If(P<Expr>, P<Block>, Option<P<Expr>>)` as today. The forms `ast::ExprKind::{IfLet, WhileLet}` will simply go poof. :)